File: Leetcode_All_Solved/1091.cpp
#include<limits.h>
#include<queue>
#include<utility>
class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid.back().size();

        vector<int> dx = {1, 1, 1, 0, 0, -1, -1, -1};
        vector<int> dy = {1, 0, -1, 1, -1, 1, 0, -1};
        
        queue<pair<int,int>> q;
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        if(grid[0][0] == 1)return (-1);
        q.push( pair<int,int>(0, 0) );
        visited[0][0] = true;

        int level = 1;
        while(q.size() != 0){
            int level_size = q.size();
            for(int l=0;l<level_size;l++){
                pair<int,int> p = q.front();
                q.pop();
                int i = p.first;
                int j = p.second;

                if(i == m-1 && j == n-1)return level;

                for(int k=0;k<dx.size();k++){
                    int x = i + dx[k];
                    int y = j + dy[k];
                    if(!(0<=x && x<m && 0<=y && y<n)) continue;
                    if(visited[x][y] || grid[x][y] == 1) continue;

                    q.push( pair<int,int>(x, y) );
                    visited[x][y] = true;
                }
            }//for l

            level++;
        }

        return (-1);
    }
};

File: Leetcode_All_Solved/1091.py
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        # edge case
        if(grid[0][0] == 1):
            return (-1)

        n = len(grid)
        di = [1, 0, -1, 0,  1, 1, -1, -1]
        dj = [0, 1, 0, -1,  1, -1, -1, 1]

        q = deque()
        ever2q = [[False for j in range(n)] for i in range(n)]
        q.append((0, 0))
        ever2q[0][0] = True

        level = 0
        while(len(q) != 0):
            level += 1

            lsize = len(q)
            for l in range(lsize):
                i, j = q[0]
                q.popleft()

                if(i==n-1 and j==n-1):
                    return level
                
                for a, b in zip(di, dj):
                    x = i+a
                    y = j+b
                    if(not(0<=x and x<n and 0<=y and y<n)):
                        continue
                    if(ever2q[x][y] == True):
                        continue

                    if(grid[x][y] != 0):
                        continue
                    
                    q.append((x, y))
                    ever2q[x][y] = True
        # end while
        return (-1)
